src/System.Collections/tests/Generic/HashSet/HashSet.Generic.Tests.cs

```diff
@@ -162,6 +162,122 @@ public void HashSet_Generic_RemoveWhere_NullMatchPredicate(int setLength)
+        {
+            var set = new HashSet<T>();
+            Assert.InRange(set.EnsureCapacity(requestedCapacity), requestedCapacity, int.MaxValue);
+        }
+
+        [Theory]
+        [InlineData(3)]
+        [InlineData(7)]
+        public void EnsureCapacity_Generic_RequestedCapacitySmallerThanCurrent_CapacityUnchanged(int currentCapacity)
+        {
+            HashSet<T> set;
+
+            // assert capacity remains the same when ensuring a capacity smaller or equal than existing
+            for (int i = 0; i <= currentCapacity; i++)
+            {
+                set = new HashSet<T>(currentCapacity);
+                Assert.Equal(currentCapacity, set.EnsureCapacity(i));
+            }
+        }
+
+        [Theory]
+        [InlineData(7)]
+        [InlineData(89)]
+        public void EnsureCapacity_Generic_ExistingCapacityRequested_SameValueReturned(int capacity)
+        {
+            var set = new HashSet<T>(capacity);
+            Assert.Equal(capacity, set.EnsureCapacity(capacity));
+
+            set = (HashSet<T>)GenericISetFactory(capacity);
+            Assert.Equal(capacity, set.EnsureCapacity(capacity));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(7)]
+        public void EnsureCapacity_Generic_EnsureCapacityCalledTwice_ReturnsSameValue(int setLength)
+        {
+            HashSet<T> set = (HashSet<T>)GenericISetFactory(setLength);
+            int capacity = set.EnsureCapacity(setLength * 2);
+            Assert.Equal(capacity, set.EnsureCapacity(setLength * 2));
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(5)]
+        [InlineData(7)]
+        [InlineData(8)]
+        public void EnsureCapacity_Generic_HashsetNotEmpty_RequestedSmallerThanCount_ReturnsAtLeastSizeOfCount(int setLength)
+        {
+            HashSet<T> set = (HashSet<T>)GenericISetFactory(setLength);
+            Assert.InRange(set.EnsureCapacity(setLength - 1), setLength, int.MaxValue);
+        }
+
+        [Theory]
+        [InlineData(7)]
+        [InlineData(20)]
+        public void EnsureCapacity_Generic_HashsetNotEmpty_SetsToAtLeastTheRequested(int setLength)
+        {
+            HashSet<T> set = (HashSet<T>)GenericISetFactory(setLength);
+
+            // get current capacity
+            int currentCapacity = set.EnsureCapacity(0);
+
+            // assert we can update to a larger capacity
+            int newCapacity = set.EnsureCapacity(currentCapacity * 2);
+            Assert.InRange(newCapacity, currentCapacity * 2, int.MaxValue);
+        }
+
+        [Fact]
+        public void EnsureCapacity_Generic_CapacityIsSetToPrimeNumberLargerOrEqualToRequested()
+        {
+            var set = new HashSet<T>();
+            Assert.Equal(17, set.EnsureCapacity(17));
+
+            set = new HashSet<T>();
+            Assert.Equal(17, set.EnsureCapacity(15));
+
+            set = new HashSet<T>();
+            Assert.Equal(17, set.EnsureCapacity(13));
```

Comment:
Why would this always return 17 if 13 is also a prime number?

Reply:
I'm guessing because the gap between the first few smaller prime numbers is small so we don't want to necessarily set to every prime number as we increase capacity.

Refer to [HashHelpers](https://github.com/dotnet/coreclr/blob/0fc8b13bfbf14a51964b95e29a67cbf53f8609c3/src/mscorlib/shared/System/Collections/HashHelpers.cs#L29).